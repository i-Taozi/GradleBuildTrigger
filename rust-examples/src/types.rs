///类型
pub fn types() {
    let decimal = 65.4321_f32;
    //不支持隐式转化
    //let integer: u8 = decimal;
    // FIXME ^ Comment out this line

    //显示转化
    let integer = decimal as u8;
    let character = integer as char;

    println!("Casting: {} -> {} -> {}", decimal, integer, character);

    // 1000 already fits in a u16
    println!("1000 as a u16 is: {}", 1000 as u16);

    // 1000 - 256 - 256 - 256 = 232
    // Under the hood, the first 8 least significant bits (LSB) are kept,
    // while the rest towards the most significant bit (MSB) get truncated.
    //println!("1000 as a u8 is : {}", 1000 as u8);
    // -1 + 256 = 255
    println!("  -1 as a u8 is : {}", (-1i8) as u8);
    println!("1000 mod 256 is : {}", 1000 % 256);

    println!(" 128 as a i16 is: {}", 128 as i16);
    //println!(" 128 as a i8 is : {}", 128 as i8);
    //println!("1000 as a u8 is : {}", 1000 as u8);
    //println!(" 232 as a i8 is : {}", 232 as i8);

    literals();
}

fn literals() {
    //类型作为后缀
    let x = 1u8;
    let y = 2u32;
    let z = 3f32;

    //非固定文字，其类型取决于如何使用
    let i = 1;
    let f = 1.0;

    //Returns the size of the pointed-to value in bytes.
    println!("size of `x` in bytes: {}", std::mem::size_of_val(&x));
    println!("size of `y` in bytes: {}", std::mem::size_of_val(&y));
    println!("size of `z` in bytes: {}", std::mem::size_of_val(&z));
    println!("size of `i` in bytes: {}", std::mem::size_of_val(&i));
    println!("size of `f` in bytes: {}", std::mem::size_of_val(&f));
}
