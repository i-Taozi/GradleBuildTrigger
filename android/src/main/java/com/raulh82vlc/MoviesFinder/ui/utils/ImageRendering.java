package com.raulh82vlc.MoviesFinder.ui.utils;

import android.widget.ImageView;

/**
 * Image Rendering contract
 */
public interface ImageRendering {
    /**
     * load an image
     *
     * @param urlPath path of the image from the API
     * @param img     ImageView where to render
     */
    void loadImage(String urlPath, ImageView img);
}
