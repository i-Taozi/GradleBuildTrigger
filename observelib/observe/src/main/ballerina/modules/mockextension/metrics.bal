// Copyright (c) 2020 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
//
// WSO2 Inc. licenses this file to you under the Apache License,
// Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import ballerina/jballerina.java;

# Holds the key and value of tag.
#
public type Tag record {
  string key;
  string value;
};

# Holds parameters required to id a metric.
#
public type MetricId record {
  string name;
  string description;
  Tag[] tags;
};

# Holds percentile and value.
#
public type PercentileValue record {
  float percentile;
  float value;
};

# Holds a time duration.
#
public type TimeWindow record {
    int seconds;
    int nanos;
};

# Holds a snapshot of statistics.
#
public type Snapshot record {
  TimeWindow timeWindow;
  float min;
  float max;
  float mean;
  float stdDev;
  PercentileValue[] percentileValues;
};

# Holds counter related metrics data.
#
public type Counter record {
  MetricId id;
  int value;
};

# Holds gauge related metrics data.
#
public type Gauge record {
  MetricId id;
  float value;
  int count;
  float sum;
  Snapshot[] snapshots;
};

# Holds polled gauge related metrics data.
#
public type PolledGauge record {
  MetricId id;
  float value;
};

# Holds metrics data.
#
public type Metrics record {
  Counter[] counters;
  Gauge[] gauges;
  PolledGauge[] polledGauges;
};

# Get all the current metrics.
#
# + return - Current metrics
public isolated function getMetrics() returns Metrics {
    json metricsJson = externGetMetrics();
    Metrics | error metrics = metricsJson.cloneWithType(Metrics);
    if (metrics is error) {
        panic error("cannot convert to Metrics record; json string: " + metricsJson.toJsonString(), metrics);
    } else {
        return metrics;
    }
}

# Get all the current metrics.
#
# + return - The metrics currently in the metrics registry as a json
isolated function externGetMetrics() returns json = @java:Method {
    name: "getMetrics",
    'class: "org.ballerinalang.observe.mockextension.MetricsUtils"
} external;
