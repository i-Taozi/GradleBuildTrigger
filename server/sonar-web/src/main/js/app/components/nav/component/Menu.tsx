/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import classNames from 'classnames';
import { LocationDescriptorObject } from 'history';
import { omit } from 'lodash';
import * as React from 'react';
import { Link, LinkProps } from 'react-router';
import Dropdown from '../../../../components/controls/Dropdown';
import Tooltip from '../../../../components/controls/Tooltip';
import { withAppState } from '../../../../components/hoc/withAppState';
import BulletListIcon from '../../../../components/icons/BulletListIcon';
import DropdownIcon from '../../../../components/icons/DropdownIcon';
import NavBarTabs from '../../../../components/ui/NavBarTabs';
import { getBranchLikeQuery, isPullRequest } from '../../../../helpers/branch-like';
import { hasMessage, translate, translateWithParameters } from '../../../../helpers/l10n';
import { getPortfolioUrl, getProjectQueryUrl } from '../../../../helpers/urls';
import { BranchLike, BranchParameters } from '../../../../types/branch-like';
import { ComponentQualifier, isPortfolioLike } from '../../../../types/component';
import './Menu.css';

const SETTINGS_URLS = [
  '/application/console',
  '/project/admin',
  '/project/baseline',
  '/project/branches',
  '/project/settings',
  '/project/quality_profiles',
  '/project/quality_gate',
  '/project/links',
  '/project_roles',
  '/project/history',
  'background_tasks',
  '/project/key',
  '/project/deletion',
  '/project/webhooks'
];

interface Props {
  appState: Pick<T.AppState, 'branchesEnabled'>;
  branchLike: BranchLike | undefined;
  branchLikes: BranchLike[] | undefined;
  component: T.Component;
  isInProgress?: boolean;
  isPending?: boolean;
  onToggleProjectInfo: () => void;
}

type Query = BranchParameters & { id: string };

export class Menu extends React.PureComponent<Props> {
  hasAnalysis = () => {
    const { branchLikes = [], component, isInProgress, isPending } = this.props;
    const hasBranches = branchLikes.length > 1;
    return hasBranches || isInProgress || isPending || component.analysisDate !== undefined;
  };

  isProject = () => {
    return this.props.component.qualifier === ComponentQualifier.Project;
  };

  isDeveloper = () => {
    return this.props.component.qualifier === ComponentQualifier.Developper;
  };

  isPortfolio = () => {
    const { qualifier } = this.props.component;
    return isPortfolioLike(qualifier);
  };

  isApplication = () => {
    return this.props.component.qualifier === ComponentQualifier.Application;
  };

  isAllChildProjectAccessible = () => {
    return Boolean(this.props.component.canBrowseAllChildProjects);
  };

  isApplicationChildInaccessble = () => {
    return this.isApplication() && !this.isAllChildProjectAccessible();
  };

  isGovernanceEnabled = () => {
    const {
      component: { extensions }
    } = this.props;

    return extensions && extensions.some(extension => extension.key.startsWith('governance/'));
  };

  getConfiguration = () => {
    return this.props.component.configuration || {};
  };

  getQuery = (): Query => {
    return { id: this.props.component.key, ...getBranchLikeQuery(this.props.branchLike) };
  };

  renderLinkWhenInaccessibleChild(label: React.ReactNode) {
    return (
      <li>
        <Tooltip
          overlay={translateWithParameters(
            'layout.all_project_must_be_accessible',
            translate('qualifier', this.props.component.qualifier)
          )}>
          <a aria-disabled="true" className="disabled-link">
            {label}
          </a>
        </Tooltip>
      </li>
    );
  }

  renderMenuLink = ({
    label,
    to,
    ...props
  }: Omit<LinkProps, 'to'> & {
    label: React.ReactNode;
    to: LocationDescriptorObject;
  }) => {
    const hasAnalysis = this.hasAnalysis();
    const isApplicationChildInaccessble = this.isApplicationChildInaccessble();
    const query = this.getQuery();
    if (isApplicationChildInaccessble) {
      return this.renderLinkWhenInaccessibleChild(label);
    }
    return (
      <li>
        {hasAnalysis ? (
          <Link to={{ ...to, query: { ...query, ...to.query } }} {...omit(props, ['to'])}>
            {label}
          </Link>
        ) : (
          <Tooltip overlay={translate('layout.must_be_configured')}>
            <a aria-disabled="true" className="disabled-link">
              {label}
            </a>
          </Tooltip>
        )}
      </li>
    );
  };

  renderDashboardLink = () => {
    const { id, ...branchLike } = this.getQuery();

    if (this.isPortfolio()) {
      return this.isGovernanceEnabled() ? (
        <li>
          <Link activeClassName="active" to={getPortfolioUrl(id)}>
            {translate('overview.page')}
          </Link>
        </li>
      ) : null;
    }

    const isApplicationChildInaccessble = this.isApplicationChildInaccessble();
    if (isApplicationChildInaccessble) {
      return this.renderLinkWhenInaccessibleChild(translate('overview.page'));
    }
    return (
      <li>
        <Link activeClassName="active" to={getProjectQueryUrl(id, branchLike)}>
          {translate('overview.page')}
        </Link>
      </li>
    );
  };

  renderCodeLink = () => {
    const isPortfolio = this.isPortfolio();
    const isApplication = this.isApplication();
    const label =
      isPortfolio || isApplication ? translate('view_projects.page') : translate('code.page');
    if (this.isDeveloper()) {
      return null;
    }

    return this.renderMenuLink({ label, to: { pathname: '/code' } });
  };

  renderActivityLink = () => {
    const { branchLike } = this.props;

    if (isPullRequest(branchLike)) {
      return null;
    }

    return this.renderMenuLink({
      label: translate('project_activity.page'),
      to: { pathname: '/project/activity' }
    });
  };

  renderIssuesLink = () => {
    return this.renderMenuLink({
      label: translate('issues.page'),
      to: { pathname: '/project/issues', query: { resolved: 'false' } }
    });
  };

  renderComponentMeasuresLink = () => {
    return this.renderMenuLink({
      label: translate('layout.measures'),
      to: { pathname: 'component_measures' }
    });
  };

  renderSecurityHotspotsLink = () => {
    const isPortfolio = this.isPortfolio();
    return (
      !isPortfolio &&
      this.renderMenuLink({
        label: translate('layout.security_hotspots'),
        to: { pathname: '/security_hotspots' }
      })
    );
  };

  renderSecurityReports = () => {
    const { branchLike, component } = this.props;
    const { extensions = [] } = component;

    if (isPullRequest(branchLike)) {
      return null;
    }

    const hasSecurityReportsEnabled = extensions.some(extension =>
      extension.key.startsWith('securityreport/')
    );

    if (!hasSecurityReportsEnabled) {
      return null;
    }

    return this.renderMenuLink({
      label: translate('layout.security_reports'),
      to: { pathname: '/project/extension/securityreport/securityreport' }
    });
  };

  renderAdministration = () => {
    const { branchLike, component } = this.props;
    const isProject = this.isProject();
    const isPortfolio = this.isPortfolio();
    const isApplication = this.isApplication();
    const query = this.getQuery();

    if (!this.getConfiguration().showSettings || isPullRequest(branchLike)) {
      return null;
    }

    const isSettingsActive = SETTINGS_URLS.some(url => window.location.href.indexOf(url) !== -1);

    const adminLinks = this.renderAdministrationLinks(query, isProject, isApplication, isPortfolio);
    if (!adminLinks.some(link => link != null)) {
      return null;
    }

    return (
      <Dropdown
        data-test="administration"
        overlay={<ul className="menu">{adminLinks}</ul>}
        tagName="li">
        {({ onToggleClick, open }) => (
          <a
            aria-expanded={open}
            aria-haspopup="menu"
            role="button"
            className={classNames('dropdown-toggle', { active: isSettingsActive || open })}
            href="#"
            id="component-navigation-admin"
            onClick={onToggleClick}>
            {hasMessage('layout.settings', component.qualifier)
              ? translate('layout.settings', component.qualifier)
              : translate('layout.settings')}
            <DropdownIcon className="little-spacer-left" />
          </a>
        )}
      </Dropdown>
    );
  };

  renderAdministrationLinks = (
    query: Query,
    isProject: boolean,
    isApplication: boolean,
    isPortfolio: boolean
  ) => {
    return [
      this.renderSettingsLink(query, isApplication, isPortfolio),
      this.renderBranchesLink(query, isProject),
      this.renderBaselineLink(query, isApplication, isPortfolio),
      this.renderConsoleAppLink(query, isApplication),
      this.renderReportSettingsLink(query, isApplication),
      ...this.renderAdminExtensions(query, isApplication),
      this.renderImportExportLink(query, isProject),
      this.renderProfilesLink(query),
      this.renderQualityGateLink(query),
      this.renderLinksLink(query),
      this.renderPermissionsLink(query),
      this.renderBackgroundTasksLink(query),
      this.renderUpdateKeyLink(query),
      this.renderWebhooksLink(query, isProject),
      this.renderDeletionLink(query)
    ];
  };

  renderProjectInformationButton = () => {
    const isProject = this.isProject();
    const isApplication = this.isApplication();
    const label = translate(isProject ? 'project' : 'application', 'info.title');
    const isApplicationChildInaccessble = this.isApplicationChildInaccessble();

    if (isPullRequest(this.props.branchLike)) {
      return null;
    }

    if (isApplicationChildInaccessble) {
      return this.renderLinkWhenInaccessibleChild(label);
    }

    return (
      (isProject || isApplication) && (
        <li>
          <a
            className="menu-button"
            onClick={(e: React.SyntheticEvent<HTMLAnchorElement>) => {
              e.preventDefault();
              e.currentTarget.blur();
              this.props.onToggleProjectInfo();
            }}
            role="button"
            tabIndex={0}>
            <BulletListIcon className="little-spacer-right" />
            {label}
          </a>
        </li>
      )
    );
  };

  renderSettingsLink = (query: Query, isApplication: boolean, isPortfolio: boolean) => {
    if (!this.getConfiguration().showSettings || isApplication || isPortfolio) {
      return null;
    }
    return (
      <li key="settings">
        <Link activeClassName="active" to={{ pathname: '/project/settings', query }}>
          {translate('project_settings.page')}
        </Link>
      </li>
    );
  };

  renderBranchesLink = (query: Query, isProject: boolean) => {
    if (
      !this.props.appState.branchesEnabled ||
      !isProject ||
      !this.getConfiguration().showSettings
    ) {
      return null;
    }

    return (
      <li key="branches">
        <Link activeClassName="active" to={{ pathname: '/project/branches', query }}>
          {translate('project_branch_pull_request.page')}
        </Link>
      </li>
    );
  };

  renderBaselineLink = (query: Query, isApplication: boolean, isPortfolio: boolean) => {
    if (!this.getConfiguration().showSettings || isApplication || isPortfolio) {
      return null;
    }
    return (
      <li key="baseline">
        <Link activeClassName="active" to={{ pathname: '/project/baseline', query }}>
          {translate('project_baseline.page')}
        </Link>
      </li>
    );
  };

  renderConsoleAppLink = (query: Query, isApplication: boolean) => {
    if (!isApplication) {
      return null;
    }
    return (
      <li key="app-console">
        <Link activeClassName="active" to={{ pathname: '/application/console', query }}>
          {translate('application_console.page')}
        </Link>
      </li>
    );
  };

  renderReportSettingsLink = (query: Query, isApplication: boolean) => {
    const extensions = this.getConfiguration().extensions || [];
    const hasGovernance = extensions.find(e => e.key === 'governance/console');

    if (!isApplication || !hasGovernance) {
      return null;
    }

    return (
      <li key="report-settings">
        <Link activeClassName="active" to={{ pathname: '/application/settings', query }}>
          {translate('application_settings.report')}
        </Link>
      </li>
    );
  };

  renderImportExportLink = (query: Query, isProject: boolean) => {
    if (!isProject) {
      return null;
    }
    return (
      <li key="import-export">
        <Link activeClassName="active" to={{ pathname: '/project/import_export', query }}>
          {translate('project_dump.page')}
        </Link>
      </li>
    );
  };

  renderProfilesLink = (query: Query) => {
    if (!this.getConfiguration().showQualityProfiles) {
      return null;
    }
    return (
      <li key="profiles">
        <Link activeClassName="active" to={{ pathname: '/project/quality_profiles', query }}>
          {translate('project_quality_profiles.page')}
        </Link>
      </li>
    );
  };

  renderQualityGateLink = (query: Query) => {
    if (!this.getConfiguration().showQualityGates) {
      return null;
    }
    return (
      <li key="quality_gate">
        <Link activeClassName="active" to={{ pathname: '/project/quality_gate', query }}>
          {translate('project_quality_gate.page')}
        </Link>
      </li>
    );
  };

  renderLinksLink = (query: Query) => {
    if (!this.getConfiguration().showLinks) {
      return null;
    }
    return (
      <li key="links">
        <Link activeClassName="active" to={{ pathname: '/project/links', query }}>
          {translate('project_links.page')}
        </Link>
      </li>
    );
  };

  renderPermissionsLink = (query: Query) => {
    if (!this.getConfiguration().showPermissions) {
      return null;
    }
    return (
      <li key="permissions">
        <Link activeClassName="active" to={{ pathname: '/project_roles', query }}>
          {translate('permissions.page')}
        </Link>
      </li>
    );
  };

  renderBackgroundTasksLink = (query: Query) => {
    if (!this.getConfiguration().showBackgroundTasks) {
      return null;
    }
    return (
      <li key="background_tasks">
        <Link activeClassName="active" to={{ pathname: '/project/background_tasks', query }}>
          {translate('background_tasks.page')}
        </Link>
      </li>
    );
  };

  renderUpdateKeyLink = (query: Query) => {
    if (!this.getConfiguration().showUpdateKey) {
      return null;
    }
    return (
      <li key="update_key">
        <Link activeClassName="active" to={{ pathname: '/project/key', query }}>
          {translate('update_key.page')}
        </Link>
      </li>
    );
  };

  renderWebhooksLink = (query: Query, isProject: boolean) => {
    if (!this.getConfiguration().showSettings || !isProject) {
      return null;
    }
    return (
      <li key="webhooks">
        <Link activeClassName="active" to={{ pathname: '/project/webhooks', query }}>
          {translate('webhooks.page')}
        </Link>
      </li>
    );
  };

  renderDeletionLink = (query: Query) => {
    const { qualifier } = this.props.component;

    if (!this.getConfiguration().showSettings) {
      return null;
    }

    if (
      ![
        ComponentQualifier.Project,
        ComponentQualifier.Portfolio,
        ComponentQualifier.Application
      ].includes(qualifier as ComponentQualifier)
    ) {
      return null;
    }

    return (
      <li key="project_delete">
        <Link activeClassName="active" to={{ pathname: '/project/deletion', query }}>
          {translate('deletion.page')}
        </Link>
      </li>
    );
  };

  renderExtension = ({ key, name }: T.Extension, isAdmin: boolean, baseQuery: Query) => {
    const pathname = isAdmin ? `/project/admin/extension/${key}` : `/project/extension/${key}`;
    const query = { ...baseQuery, qualifier: this.props.component.qualifier };
    return (
      <li key={key}>
        <Link activeClassName="active" to={{ pathname, query }}>
          {name}
        </Link>
      </li>
    );
  };

  renderAdminExtensions = (query: Query, isApplication: boolean) => {
    const extensions = this.getConfiguration().extensions || [];
    return extensions
      .filter(e => !isApplication || e.key !== 'governance/console')
      .map(e => this.renderExtension(e, true, query));
  };

  renderExtensions = () => {
    const query = this.getQuery();
    const extensions = this.props.component.extensions || [];
    const withoutSecurityExtension = extensions.filter(
      extension =>
        !extension.key.startsWith('securityreport/') && !extension.key.startsWith('governance/')
    );

    if (withoutSecurityExtension.length === 0) {
      return null;
    }

    return (
      <Dropdown
        data-test="extensions"
        overlay={
          <ul className="menu">
            {withoutSecurityExtension.map(e => this.renderExtension(e, false, query))}
          </ul>
        }
        tagName="li">
        {({ onToggleClick, open }) => (
          <a
            aria-expanded={open}
            aria-haspopup="menu"
            role="button"
            className={classNames('dropdown-toggle', { active: open })}
            href="#"
            id="component-navigation-more"
            onClick={onToggleClick}>
            {translate('more')}
            <DropdownIcon className="little-spacer-left" />
          </a>
        )}
      </Dropdown>
    );
  };

  render() {
    return (
      <div className="display-flex-center display-flex-space-between">
        <NavBarTabs>
          {this.renderDashboardLink()}
          {this.renderIssuesLink()}
          {this.renderSecurityHotspotsLink()}
          {this.renderSecurityReports()}
          {this.renderComponentMeasuresLink()}
          {this.renderCodeLink()}
          {this.renderActivityLink()}
          {this.renderExtensions()}
        </NavBarTabs>
        <NavBarTabs>
          {this.renderAdministration()}
          {this.renderProjectInformationButton()}
        </NavBarTabs>
      </div>
    );
  }
}

export default withAppState(Menu);
