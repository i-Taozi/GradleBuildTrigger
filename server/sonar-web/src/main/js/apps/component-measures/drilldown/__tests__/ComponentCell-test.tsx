/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { Link } from 'react-router';
import {
  mockComponentMeasure,
  mockComponentMeasureEnhanced
} from '../../../../helpers/mocks/component';
import { mockMetric } from '../../../../helpers/testMocks';
import { ComponentQualifier } from '../../../../types/component';
import { MetricKey } from '../../../../types/metrics';
import { enhanceComponent } from '../../utils';
import ComponentCell, { ComponentCellProps } from '../ComponentCell';

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('default');
});

it.each([
  [ComponentQualifier.Project, undefined],
  [ComponentQualifier.Project, 'develop'],
  [ComponentQualifier.Application, undefined],
  [ComponentQualifier.Application, 'develop'],
  [ComponentQualifier.Portfolio, undefined],
  [ComponentQualifier.Portfolio, 'develop']
])(
  'should render correctly for a "%s" root component and a component with branch "%s"',
  (rootComponentQualifier: ComponentQualifier, componentBranch: string | undefined) => {
    expect(
      shallowRender({
        rootComponent: mockComponentMeasure(false, { qualifier: rootComponentQualifier }),
        component: mockComponentMeasureEnhanced({ branch: componentBranch })
      })
    ).toMatchSnapshot();
  }
);

it('should properly deal with key and refKey', () => {
  expect(
    shallowRender({
      component: mockComponentMeasureEnhanced({
        qualifier: ComponentQualifier.SubPortfolio,
        refKey: 'port-key'
      })
    })
      .find(Link)
      .props().to
  ).toEqual(expect.objectContaining({ query: expect.objectContaining({ id: 'port-key' }) }));

  expect(
    shallowRender()
      .find(Link)
      .props().to
  ).toEqual(
    expect.objectContaining({
      query: expect.objectContaining({ id: 'foo', selected: 'foo:src/index.tsx' })
    })
  );
});

it.each([
  [
    ComponentQualifier.File,
    MetricKey.bugs,
    expect.objectContaining({
      pathname: '/component_measures',
      query: expect.objectContaining({ branch: 'develop' })
    })
  ],
  [
    ComponentQualifier.Directory,
    MetricKey.bugs,
    expect.objectContaining({
      pathname: '/component_measures',
      query: expect.objectContaining({ branch: 'develop' })
    })
  ],
  [
    ComponentQualifier.Project,
    MetricKey.projects,
    expect.objectContaining({
      pathname: '/dashboard',
      query: expect.objectContaining({ branch: 'develop' })
    })
  ],
  [
    ComponentQualifier.Application,
    MetricKey.releasability_rating,
    expect.objectContaining({
      pathname: '/dashboard',
      query: expect.objectContaining({ branch: 'develop' })
    })
  ],
  [
    ComponentQualifier.Project,
    MetricKey.releasability_rating,
    expect.objectContaining({
      pathname: '/dashboard',
      query: expect.objectContaining({ branch: 'develop' })
    })
  ],
  [
    ComponentQualifier.Application,
    MetricKey.alert_status,
    expect.objectContaining({
      pathname: '/dashboard',
      query: expect.objectContaining({ branch: 'develop' })
    })
  ],
  [
    ComponentQualifier.Project,
    MetricKey.alert_status,
    expect.objectContaining({
      pathname: '/dashboard',
      query: expect.objectContaining({ branch: 'develop' })
    })
  ]
])(
  'should display the proper link path for %s component qualifier and %s metric key',
  (componentQualifier: ComponentQualifier, metricKey: MetricKey, expectedTo: any) => {
    const wrapper = shallowRender(
      {
        component: mockComponentMeasureEnhanced({
          qualifier: componentQualifier,
          branch: 'develop'
        })
      },
      metricKey
    );

    expect(wrapper.find(Link).props().to).toEqual(expectedTo);
  }
);

function shallowRender(overrides: Partial<ComponentCellProps> = {}, metricKey = MetricKey.bugs) {
  const metric = mockMetric({ key: metricKey });
  const component = enhanceComponent(
    mockComponentMeasure(true, {
      measures: [{ metric: metric.key, value: '1', bestValue: false }]
    }),
    metric,
    { [metric.key]: metric }
  );

  return shallow<ComponentCellProps>(
    <ComponentCell
      component={component}
      metric={metric}
      rootComponent={mockComponentMeasure(false)}
      view="list"
      {...overrides}
    />
  );
}
