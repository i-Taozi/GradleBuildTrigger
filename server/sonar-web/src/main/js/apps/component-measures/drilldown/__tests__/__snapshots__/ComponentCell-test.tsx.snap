// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly for a "APP" root component and a component with branch "develop" 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "branch": "develop",
            "id": "foo",
            "metric": "bugs",
            "selected": "foo",
            "view": "list",
          },
        }
      }
    >
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="TRK"
        />
        <span>
          Foo
        </span>
        <BranchIcon
          className="spacer-left little-spacer-right"
        />
        <span
          className="note"
        >
          develop
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly for a "APP" root component and a component with branch "undefined" 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "id": "foo",
            "metric": "bugs",
            "selected": "foo",
            "view": "list",
          },
        }
      }
    >
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="TRK"
        />
        <span>
          Foo
        </span>
        <span
          className="spacer-left badge"
        >
          branches.main_branch
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly for a "TRK" root component and a component with branch "develop" 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "branch": "develop",
            "id": "foo",
            "metric": "bugs",
            "selected": "foo",
            "view": "list",
          },
        }
      }
    >
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="TRK"
        />
        <span>
          Foo
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly for a "TRK" root component and a component with branch "undefined" 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "id": "foo",
            "metric": "bugs",
            "selected": "foo",
            "view": "list",
          },
        }
      }
    >
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="TRK"
        />
        <span>
          Foo
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly for a "VW" root component and a component with branch "develop" 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "branch": "develop",
            "id": "foo",
            "metric": "bugs",
            "selected": "foo",
            "view": "list",
          },
        }
      }
    >
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="TRK"
        />
        <span>
          Foo
        </span>
        <BranchIcon
          className="spacer-left little-spacer-right"
        />
        <span
          className="note"
        >
          develop
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly for a "VW" root component and a component with branch "undefined" 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "id": "foo",
            "metric": "bugs",
            "selected": "foo",
            "view": "list",
          },
        }
      }
    >
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="TRK"
        />
        <span>
          Foo
        </span>
        <span
          className="spacer-left badge"
        >
          branches.main_branch
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly: default 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo:src/index.tsx"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "id": "foo",
            "metric": "bugs",
            "selected": "foo:src/index.tsx",
            "view": "list",
          },
        }
      }
    >
      <span
        title="foo:src/index.tsx"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="FIL"
        />
        <span
          className="note"
        >
          src
          /
        </span>
        <span>
          index.tsx
        </span>
      </span>
    </Link>
  </div>
</td>
`;
